File: crates/transport/rrd-client/tests/sdk_conformance.rs
use rrd_client::{is_unauthenticated, ClientConfig, Error, RequestOptions, RrdClient};
use rrd_contract::{
    FollowChangefeed, ReadChangefeed, SdkConformanceCorpus, SessionEndState, TransactionState,
};
use rrd_core::digest;
use serde::Deserialize;
use std::collections::BTreeMap;
use std::net::SocketAddr;
use std::path::PathBuf;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
struct HarnessManifest {
    format_version: u16,
    corpus_sha256: String,
    corpus_path: PathBuf,
    base_url: String,
    retry_base_urls: BTreeMap<String, String>,
    incompatible_version_url: String,
    shutdown_path: PathBuf,
}

fn address(url: &str) -> SocketAddr {
    url.strip_prefix("http://")
        .expect("conformance URL is loopback HTTP")
        .trim_end_matches('/')
        .parse()
        .expect("conformance URL contains a socket address")
}

fn mutation(step: &str) -> RequestOptions {
    RequestOptions::mutation(
        &format!("rust-{step}-request"),
        &format!("rust-{step}-operation"),
        &format!("rust-{step}-key"),
    )
    .unwrap()
}

fn read(step: &str) -> RequestOptions {
    RequestOptions::read(
        &format!("rust-{step}-request"),
        &format!("rust-{step}-operation"),
    )
    .unwrap()
}

#[tokio::test]
async fn rust_sdk_passes_shared_real_daemon_conformance() {
    let Ok(manifest_path) = std::env::var("RRD_SDK_CONFORMANCE_MANIFEST") else {
        return;
    };
    let manifest_bytes = std::fs::read(manifest_path).unwrap();
    let manifest: HarnessManifest = serde_json::from_slice(&manifest_bytes).unwrap();
    assert_eq!(manifest.format_version, 1);
    assert!(!manifest.shutdown_path.as_os_str().is_empty());
    let corpus_bytes = std::fs::read(&manifest.corpus_path).unwrap();
    assert_eq!(manifest.corpus_sha256, digest::sha256_hex(&corpus_bytes));
    let corpus: SdkConformanceCorpus = serde_json::from_slice(&corpus_bytes).unwrap();
    corpus.validate().unwrap();

    let retry_client = RrdClient::connect_local(
        address(&manifest.retry_base_urls["rust"]),
        corpus.identity.instance.clone(),
        ClientConfig {
            request_timeout: Duration::from_secs(5),
            max_attempts: corpus.expected.retry_attempts,
        },
    )
    .unwrap();
    assert_eq!(
        retry_client.capabilities().await.unwrap().protocol_version,
        corpus.protocol_version
    );

    let incompatible = RrdClient::connect_local(
        address(&manifest.incompatible_version_url),
        corpus.identity.instance.clone(),
        ClientConfig::default(),
    )
    .unwrap()
    .capabilities()
    .await
    .unwrap_err();
    assert!(matches!(
        incompatible,
        Error::UnsupportedProtocol { version, .. }
            if version == corpus.incompatible_protocol_version
    ));

    let client = RrdClient::connect_local(
        address(&manifest.base_url),
        corpus.identity.instance.clone(),
        ClientConfig {
            request_timeout: Duration::from_secs(10),
            max_attempts: 2,
        },
    )
    .unwrap();
    let capabilities = client.capabilities().await.unwrap();
    assert_eq!(capabilities.protocol, corpus.protocol);
    assert_eq!(
        client.endpoint_catalogue().await.unwrap().endpoints.len(),
        usize::from(corpus.expected.endpoint_count)
    );

    let wrong_key = client
        .create_session(
            corpus.identity.principal.clone(),
            "wrong-sdk-conformance-key",
            corpus.session.create.clone(),
            mutation("wrong-key"),
        )
        .await
        .unwrap_err();
    assert!(is_unauthenticated(&wrong_key));

    let session = client
        .create_session(
            corpus.identity.principal.clone(),
            &corpus.identity.api_key,
            corpus.session.create.clone(),
            mutation("session-create"),
        )
        .await
        .unwrap();
    let session = client
        .renew_session(
            &session,
            corpus.session.renew.clone(),
            mutation("session-renew"),
        )
        .await
        .unwrap();

    client
        .ensure_vector_collection(
            &session,
            corpus.vector.ensure.clone(),
            mutation("vector-ensure"),
        )
        .await
        .unwrap();

    let preview_transaction = client
        .begin_transaction(
            &session,
            corpus.transaction.preview_begin.clone(),
            mutation("preview-begin"),
        )
        .await
        .unwrap();
    let preview = client
        .preview_transaction(
            &session,
            &preview_transaction.transaction_id,
            corpus.transaction.preview.clone(),
            mutation("transaction-preview"),
        )
        .await
        .unwrap();
    assert_eq!(preview.transaction_id, preview_transaction.transaction_id);
    let aborted = client
        .abort_transaction(
            &session,
            &preview_transaction.transaction_id,
            corpus.transaction.abort.clone(),
            mutation("transaction-abort"),
        )
        .await
        .unwrap();
    assert_eq!(aborted.state, TransactionState::Aborted);

    let commit_transaction = client
        .begin_transaction(
            &session,
            corpus.transaction.commit_begin.clone(),
            mutation("commit-begin"),
        )
        .await
        .unwrap();
    let committed = client
        .commit_transaction(
            &session,
            &commit_transaction.transaction_id,
            corpus.transaction.commit.clone(),
            RequestOptions::new(
                "rust-transaction-commit-request",
                "rust-transaction-commit-operation",
                Some("rust-transaction-commit-key"),
                Some(
                    u64::try_from(
                        SystemTime::now()
                            .duration_since(UNIX_EPOCH)
                            .unwrap()
                            .as_millis(),
                    )
                    .unwrap()
                    .saturating_add(corpus.transaction.commit_deadline_timeout_ms),
                ),
            )
            .unwrap(),
        )
        .await
        .unwrap();
    assert_eq!(
        committed.operation_sha256,
        corpus.transaction.commit.operation_sha256
    );

    let query = client
        .execute_query(&session, corpus.query.clone(), read("query"))
        .await
        .unwrap();
    assert!(query
        .rows
        .iter()
        .any(|row| row.identity == corpus.expected.query_identity));
    let vectors = client
        .search_vectors(
            &session,
            corpus.vector.search.clone(),
            read("vector-search"),
        )
        .await
        .unwrap();
    assert!(vectors.hits.iter().any(|hit| {
        format!("{}:{}", hit.reference.kind, hit.reference.id) == corpus.expected.vector_reference
    }));

    let changes = client
        .read_changefeed(
            &session,
            corpus.changefeed.read.clone(),
            read("changefeed-read"),
        )
        .await
        .unwrap();
    let follow_read = ReadChangefeed {
        after_cursor: changes.head_cursor,
        ..corpus.changefeed.read.clone()
    };
    let followed = client
        .follow_changefeed(
            &session,
            FollowChangefeed {
                read: follow_read.clone(),
                wait_timeout_ms: corpus.changefeed.follow_wait_timeout_ms,
            },
            read("changefeed-follow"),
        )
        .await
        .unwrap();
    assert!(followed.timed_out);
    let cancelled = tokio::time::timeout(
        Duration::from_millis(corpus.changefeed.cancel_after_ms),
        client.follow_changefeed(
            &session,
            FollowChangefeed {
                read: follow_read,
                wait_timeout_ms: corpus.changefeed.cancellation_wait_timeout_ms,
            },
            read("changefeed-cancel"),
        ),
    )
    .await;
    assert!(cancelled.is_err());

    let backup = client
        .create_backup(
            &session,
            corpus.backup.create.clone(),
            mutation("backup-create"),
        )
        .await
        .unwrap();
    let backup_prefix = format!("{}--", corpus.backup.create.label);
    let backup_suffix = backup.backup.label.strip_prefix(&backup_prefix).unwrap();
    assert_eq!(backup_suffix.len(), 16);
    assert!(backup_suffix.bytes().all(|byte| byte.is_ascii_hexdigit()));
    let backups = client
        .list_backups(&session, corpus.backup.list.clone(), read("backup-list"))
        .await
        .unwrap();
    assert!(backups
        .backups
        .iter()
        .any(|entry| entry.backup_sha256 == backup.backup.backup_sha256));
    let estate = client
        .read_estate(
            &session,
            corpus.identity.estate.clone(),
            corpus.estate.clone(),
            read("estate-read"),
        )
        .await
        .unwrap();
    assert_eq!(estate.revision, corpus.expected.estate_revision);

    let closed = client
        .close_session(&session, corpus.session.close, mutation("session-close"))
        .await
        .unwrap();
    assert_eq!(closed.state, SessionEndState::Closed);
    eprintln!(
        "SDK conformance OK: language=rust corpus_sha256={}",
        manifest.corpus_sha256
    );
}

File: sdks/typescript/tests/sdk-conformance.ts
import assert from "node:assert/strict";
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import {
  type RequestOptions,
  type RequestPayload,
  RrdApiError,
  RrdClient,
  RrdClientError,
  type Session,
} from "../src/index.js";

type JsonObject = Record<string, unknown>;

interface HarnessManifest {
  format_version: number;
  corpus_sha256: string;
  corpus_path: string;
  base_url: string;
  retry_base_urls: Record<string, string>;
  incompatible_version_url: string;
}

interface Corpus {
  format_version: number;
  protocol: string;
  protocol_version: number;
  incompatible_protocol_version: number;
  required_domains: string[];
  identity: { instance: string; estate: string; principal: string; api_key: string };
  session: { create: JsonObject; renew: JsonObject; close: JsonObject };
  transaction: {
    preview_begin: JsonObject;
    preview: JsonObject;
    abort: JsonObject;
    commit_begin: JsonObject;
    commit_deadline_timeout_ms: number;
    commit: JsonObject & { operation_sha256: string };
  };
  query: JsonObject;
  vector: { ensure: JsonObject; search: JsonObject };
  changefeed: {
    read: JsonObject & { after_cursor: number; limit: number; scope: string };
    follow_wait_timeout_ms: number;
    cancellation_wait_timeout_ms: number;
    cancel_after_ms: number;
  };
  backup: { create: JsonObject & { label: string }; list: JsonObject };
  estate: JsonObject;
  expected: {
    endpoint_count: number;
    query_identity: string;
    vector_reference: string;
    estate_revision: number;
    typed_error: string;
    retry_attempts: number;
  };
}

const manifestPath = process.env.RRD_SDK_CONFORMANCE_MANIFEST;
assert.ok(manifestPath, "RRD_SDK_CONFORMANCE_MANIFEST is required");
const manifest = JSON.parse(readFileSync(manifestPath, "utf8")) as HarnessManifest;
const corpusBytes = readFileSync(manifest.corpus_path);
const corpus = JSON.parse(corpusBytes.toString("utf8")) as Corpus;
assert.equal(manifest.format_version, corpus.format_version);
assert.equal(createHash("sha256").update(corpusBytes).digest("hex"), manifest.corpus_sha256);
assert.deepEqual(
  new Set(corpus.required_domains),
  new Set([
    "auth",
    "backup",
    "cancellation",
    "crud",
    "estate",
    "live_feeds",
    "query",
    "retries",
    "sessions",
    "transactions",
    "typed_errors",
    "vectors",
    "versions",
  ]),
);

const mutation = (
  step: string,
  session?: Session,
  pathParameters?: Record<string, string>,
): RequestOptions => ({
  requestId: `typescript-${step}-request`,
  operationId: `typescript-${step}-operation`,
  idempotencyKey: `typescript-${step}-key`,
  ...(session ? { session } : {}),
  ...(pathParameters ? { pathParameters } : {}),
});
const read = (
  step: string,
  session: Session,
  pathParameters?: Record<string, string>,
): RequestOptions => ({
  requestId: `typescript-${step}-request`,
  operationId: `typescript-${step}-operation`,
  session,
  ...(pathParameters ? { pathParameters } : {}),
});

const retryBaseUrl = manifest.retry_base_urls.typescript;
assert.ok(retryBaseUrl, "TypeScript retry endpoint is required");
const retryClient = new RrdClient({
  baseUrl: retryBaseUrl,
  instance: corpus.identity.instance,
  maxAttempts: corpus.expected.retry_attempts,
});
assert.equal((await retryClient.capabilities()).protocol_version, corpus.protocol_version);

const incompatibleClient = new RrdClient({
  baseUrl: manifest.incompatible_version_url,
  instance: corpus.identity.instance,
});
await assert.rejects(incompatibleClient.capabilities(), RrdClientError);

const client = new RrdClient({ baseUrl: manifest.base_url, instance: corpus.identity.instance });
assert.equal((await client.capabilities()).protocol, corpus.protocol);
assert.equal((await client.endpointCatalogue()).endpoints.length, corpus.expected.endpoint_count);

await assert.rejects(
  client.createSession(
    corpus.identity.principal,
    "wrong-sdk-conformance-key",
    corpus.session.create as RequestPayload<"session-create">,
    mutation("wrong-key"),
  ),
  (error) => error instanceof RrdApiError && error.code === corpus.expected.typed_error,
);

let session = await client.createSession(
  corpus.identity.principal,
  corpus.identity.api_key,
  corpus.session.create as RequestPayload<"session-create">,
  mutation("session-create"),
);
const renewed = await client.call(
  "session-renew",
  corpus.session.renew as RequestPayload<"session-renew">,
  mutation("session-renew", session, { session: session.lease.session_id }),
);
session = { principalId: session.principalId, lease: renewed };

await client.call(
  "vector-collection-ensure",
  corpus.vector.ensure as RequestPayload<"vector-collection-ensure">,
  mutation("vector-ensure", session),
);
const previewLease = await client.call(
  "transaction-begin",
  corpus.transaction.preview_begin as RequestPayload<"transaction-begin">,
  mutation("preview-begin", session),
);
const preview = await client.call(
  "transaction-preview",
  corpus.transaction.preview as RequestPayload<"transaction-preview">,
  mutation("transaction-preview", session, { transaction: previewLease.transaction_id }),
);
assert.equal(preview.transaction_id, previewLease.transaction_id);
const aborted = await client.call(
  "transaction-abort",
  corpus.transaction.abort as RequestPayload<"transaction-abort">,
  mutation("transaction-abort", session, { transaction: previewLease.transaction_id }),
);
assert.equal(aborted.state, "aborted");

const commitLease = await client.call(
  "transaction-begin",
  corpus.transaction.commit_begin as RequestPayload<"transaction-begin">,
  mutation("commit-begin", session),
);
const committed = await client.call(
  "transaction-commit",
  corpus.transaction.commit as RequestPayload<"transaction-commit">,
  {
    ...mutation("transaction-commit", session, { transaction: commitLease.transaction_id }),
    deadlineUnixMs: Date.now() + corpus.transaction.commit_deadline_timeout_ms,
  },
);
assert.equal(committed.operation_sha256, corpus.transaction.commit.operation_sha256);

const query = await client.call(
  "query-execute",
  corpus.query as RequestPayload<"query-execute">,
  read("query", session),
);
assert.ok(query.rows.some((row) => row.identity === corpus.expected.query_identity));
const vectors = await client.call(
  "vector-search",
  corpus.vector.search as RequestPayload<"vector-search">,
  read("vector-search", session),
);
assert.ok(
  vectors.hits.some(
    (hit) => `${hit.reference.kind}:${hit.reference.id}` === corpus.expected.vector_reference,
  ),
);

const changes = await client.call(
  "changefeed-read",
  corpus.changefeed.read as RequestPayload<"changefeed-read">,
  read("changefeed-read", session),
);
const followRead = { ...corpus.changefeed.read, after_cursor: changes.head_cursor };
const followed = await client.call(
  "changefeed-follow",
  { read: followRead, wait_timeout_ms: corpus.changefeed.follow_wait_timeout_ms },
  read("changefeed-follow", session),
);
assert.equal(followed.timed_out, true);
const controller = new AbortController();
const cancellation = setTimeout(() => controller.abort(), corpus.changefeed.cancel_after_ms);
try {
  await assert.rejects(
    client.call(
      "changefeed-follow",
      { read: followRead, wait_timeout_ms: corpus.changefeed.cancellation_wait_timeout_ms },
      { ...read("changefeed-cancel", session), signal: controller.signal },
    ),
    RrdClientError,
  );
} finally {
  clearTimeout(cancellation);
}

const backup = await client.call(
  "backup-create",
  corpus.backup.create as RequestPayload<"backup-create">,
  mutation("backup-create", session),
);
const backupPrefix = `${corpus.backup.create.label}--`;
assert.match(backup.backup.label.slice(backupPrefix.length), /^[0-9a-f]{16}$/);
const backups = await client.call(
  "backup-list",
  corpus.backup.list as RequestPayload<"backup-list">,
  read("backup-list", session),
);
assert.ok(backups.backups.some((entry) => entry.backup_sha256 === backup.backup.backup_sha256));
const estate = await client.call(
  "estate-read",
  corpus.estate as RequestPayload<"estate-read">,
  read("estate-read", session, { estate: corpus.identity.estate }),
);
assert.equal(estate.revision, corpus.expected.estate_revision);
const closed = await client.call(
  "session-close",
  corpus.session.close as RequestPayload<"session-close">,
  mutation("session-close", session, { session: session.lease.session_id }),
);
assert.equal(closed.state, "closed");
console.error(`SDK conformance OK: language=typescript corpus_sha256=${manifest.corpus_sha256}`);
